book/src/chart_exercise.md

```diff
@@ -0,0 +1,102 @@
+# Chart Exercise
+
+Next, we'll begin manipulating strings and primitive arrays. You'll write JNI
+code to wrap a Chart Plotting library - which, along the way, copies primitive
+arrays and Strings, and return a new Java String.
+
+## Goal
+
+The target Java usage appears as:
+
+```Java
+package jni_rs_book;
+
+public class ChartRenderer {
+
+    public String renderChart(/* Various parameters for the drawing the chart */) {
+    
+    }
+}
+```
+
+And the Rust function that you should figure out how to wrap is:
+
+```rust,noplaypen
+/// Given various data, render a Chart as a String. 
+/// Upon failure, return Error.
+fn render_chart(
+    width: u32,
+    height: u32,
+    x_label: &str,
+    y_label: &str,
+    x_start: f64,
+    x_end: f64,
+    y_start: f64,
+    y_end: f64,
+    data: Vec<(f64, f64)>,
+) -> Result<String, anyhow::Error> {
+    use plotlib::page::Page;
+    use plotlib::repr::Plot;
+    use plotlib::style::{PointMarker, PointStyle};
+    use plotlib::view::ContinuousView;
+
+    let plot = Plot::new(data).point_style(PointStyle::new().marker(PointMarker::Circle));
+
+    let v = ContinuousView::new()
+        .add(plot)
+        .x_range(x_start, x_end)
+        .y_range(y_start, y_end)
+        .x_label(x_label)
+        .y_label(y_label);
+
+    Ok(Page::single(&v)
+        .dimensions(width, height)
+        .to_text()
+        .map_err(|err| err.compat())?)
+}
+```
+
+If you get stuck, try working with the docs, or the next page for hints, or the
+next-next page for a solution.
+
+## Docs to get you started
+These are some selections from the `jni-rs` docs that may help you.
+
+### Reading Java Array Lengths
+[`JNIEnv::get_array_length(&self, array: jarray) ->
+Result<jsize>`](https://docs.rs/jni/0.18.0/jni/struct.JNIEnv.html#method.get_array_length)
+can be used to read the length of Java Arrays. Note that it can be better to
+pass the array length together with the array, down to the native layer, when
+that is an option, since it saves an upcall.
```

Comment:
That's a good point, it's very likely that some JNI calls are less expensive than others, and that getting length is common enough that some VMs will handle it relatively well. I'll remove that suggestion, because it depends on the implementation.

Reply:
FYI, there are some benches that could be extended. Mostly they cover native -> Java method invocation using approaches on the various level of abstraction (from "pass class and method names as strings" to "look up the class ref and method ID and reuse between repeated invocations").

Reply:
https://github.com/jni-rs/jni-rs/blob/master/benches/api_calls.rs

Reply:
I'm tempted to include a reference to the benches in the materials (perhaps in future topics). What do you think?

Reply:
Agree, the set of tests already gives insights into the price of some features or coding patterns; it could be extended with more tests. It's good for users to understand the trade-offs of various options we provide, on their JVM version.